### PR TITLE
Desktop & CliClient: bugfix for several images download problem in rest Api

### PR DESCRIPTION
**Pull request by joybinchen**
Images with weird file ext in URL will not show after web clipping. 
Try to get the correct ext from HTTP response content-type.

test case
> https://wikimedia.org/api/rest_v1/media/math/render/svg/07808adf2ca4b2aeb69b6c15fb6251a3a3617c91

in Page
> https://en.wikipedia.org/wiki/Formula

### CODE REVIEW

File: ReactNativeClient/lib/services/rest/Api.js [hunk 1]
```diff
@@ -433,7 +433,27 @@ class Api {
 			if (isDataUrl) {
 				await shim.imageFromDataUrl(url, imagePath);
 			} else {
-				await shim.fetchBlob(url, { path: imagePath });
+				const response = await shim.fetchBlob(url, { path: imagePath });
+				let content_type = response.headers['content-type'];
```
Comment (laurent22) [inline ReactNativeClient/lib/services/rest/Api.js]: Please use camel case for variable names (contentType).

File: ReactNativeClient/lib/services/rest/Api.js [hunk 2]
```diff
@@ -433,7 +433,27 @@ class Api {
...
+				if (content_type) {
+					content_type = content_type.split(';')[0]
+				}
+
+				const imageMimeExtensions = {
```
Comment (laurent22) [inline ReactNativeClient/lib/services/rest/Api.js]: Let's re-use Resource.isSupportedImageMimeType for the list of mime types, and mimeUtils.toFileExtension to get the extension.

File: ReactNativeClient/lib/services/rest/Api.js [hunk 3]
```diff
@@ -433,7 +433,27 @@ class Api {
...
+					'image/jpg':	'.jpg',
+					'image/gif':	'.gif',
+					'image/png':	'.png',
+					'image/svg+xml':'.svg',
+					'image/webp':	'.webp',
+				};
+				let normalExt = imageMimeExtensions[content_type];
+				if (normalExt && normalExt !== fileExt) {
+					let normalPath = imagePath.slice(0, -fileExt.length) + normalExt;
+					shim.fsDriver().move(imagePath, normalPath);
```
Comment (laurent22) [inline ReactNativeClient/lib/services/rest/Api.js]: move() is async so there should be an "await" here

**Comment by laurent22** [comment]
Thanks @joybinchen, I've added a few comments to the PR. Would you be able to add test units for this (in services_rest_Api) or would that be too complex? (not sure how to mock the download and content-type in particular).

**Comment by joybinchen** [comment]
Still something wrong.
Using await shim.fsDriver().exists(imagePath) is unable to detect different images in a single page that maps to the same target imagePath. 
The problem happens when the image URLs are sharing a very long common tail. 
for example:

2019-01-19 20:01:52: "fetchBlob", "https://img.huxiucdn.com/article/content/201901/18/222644743133.jpg?imageView2/2/w/1000/format/jpg/interlace/1/q/85", "/home/joybin/.config/joplin-desktop/tmp/85.jpgimageview22w1000formatjpginterlace1q85"
2019-01-19 20:01:52: "fetchBlob", "https://img.huxiucdn.com/article/content/201901/18/222644926194.jpg?imageView2/2/w/1000/format/jpg/interlace/1/q/85", "/home/joybin/.config/joplin-desktop/tmp/85.jpgimageview22w1000formatjpginterlace1q85"
2019-01-19 20:01:52: "fetchBlob", "https://img.huxiucdn.com/article/content/201901/18/222645029954.jpg?imageView2/2/w/1000/format/jpg/interlace/1/q/85", "/home/joybin/.config/joplin-desktop/tmp/85.jpgimageview22w1000formatjpginterlace1q85"
2019-01-19 20:01:52: "fetchBlob", "https://img.huxiucdn.com/article/content/201901/18/222652888460.jpg?imageView2/2/w/1000/format/jpg/interlace/1/q/85", "/home/joybin/.config/joplin-desktop/tmp/85.jpgimageview22w1000formatjpginterlace1q85"

**Comment by joybinchen** [comment]
It's better to use memory install of file to pass the image content around. 
Or unconditionally use an random filename & meaningful ext as the intermediate image path.

**Comment by joybinchen** [comment]
Problem solved, and all test passed expect for the StringUtils test case which didn't pass on master branch , neither.
This PR contain 3 bug fix and one enhancement:
1. shim.resizeImage_ use sharp.resize for all images, and make it hard to detect image content change in jasmine.
2. downloadImages_ will conflict on urls with common tail in the same page, and make image test result unstable.
3. downloadImages_ doesn't get correct mime and ext for svg image with weird url tail
4. Resource.title doesn't contain corrent filename from http response.

**Comment by laurent22** [comment]
There's a lot of code in this pull request so it's difficult to review it (I know that's it's my fault as I've asked for tests, which are not easy at all to provide). But I'm wondering - why is so much code needed to detect the type of an image? Can't we ignore all the http stuff, and simply read the file header - it should tell us whether it's jpeg, png, etc. There's a package [here](https://www.npmjs.com/package/image-type) to do this. What do you think?

**Comment by joybinchen** [comment]
The PR is complex because it try to fix 4 related problem. I have made each commit specifically for one of the problem in my branch, step by step. but in this PR, them were totally mixed up. Shall I separate them into 4 PR?

**Comment by laurent22** [comment]
Sorry the late reply @joybinchen. Yes please if you could submit one PR for the first issue, then we review it, and then submit the second PR, etc. especially if they are all a bit related. That way it's easier to review and discuss the pro and con of each approach. So I'll close this PR and feel free to submit the first PR.

**Pull request closed without merging**